src/Futhark/LSP/Tool.hs

```diff
@@ -54,10 +54,16 @@ findDefinitionRange state (Just path) l c = do
 findDefinitionRange _ _ _ _ = Nothing
 
 queryAtPos :: State -> Pos -> Maybe AtPos
-queryAtPos state pos =
-  case stateProgram state of
-    Nothing -> Nothing
-    Just loaded_prog -> atPos (lpImports loaded_prog) pos
+queryAtPos state pos = do
```

Comment:
Need to get the current file content via [getVirtualFile](https://hackage.haskell.org/package/lsp-1.4.0.0/docs/Language-LSP-Server.html#v:getVirtualFile) here, not sure how to do that except by adding `LspM () ...` to the return type of `queryAtPos`, `findDefinitionRange` and `getHoverInfoFromState`. But feels wrong to do that, any tips? @athas

Reply:
Perhaps you can call that function outermost and pass on the result as an extra parameter.